crates/reth/exex/src/state_diff_exex.rs

```diff
@@ -0,0 +1,75 @@
+use std::sync::Arc;
+
+use alloy_rpc_types::BlockNumHash;
+use alpen_reth_db::StateDiffStore;
+use futures_util::TryStreamExt;
+use reth_exex::{ExExContext, ExExEvent};
+use reth_node_api::{FullNodeComponents, NodeTypes};
+use reth_primitives::EthPrimitives;
+use reth_provider::Chain;
+use tracing::{debug, error};
+
+pub struct StateDiffGenerator<
+    Node: FullNodeComponents<Types: NodeTypes<Primitives = EthPrimitives>>,
+    S: StateDiffStore + Clone,
+> {
+    ctx: ExExContext<Node>,
+    db: Arc<S>,
+}
+
+impl<
+        Node: FullNodeComponents<Types: NodeTypes<Primitives = EthPrimitives>>,
+        S: StateDiffStore + Clone,
+    > StateDiffGenerator<Node, S>
+{
+    pub fn new(ctx: ExExContext<Node>, db: Arc<S>) -> Self {
+        Self { ctx, db }
+    }
+
+    fn commit(&mut self, chain: &Chain) -> eyre::Result<Option<BlockNumHash>> {
+        let mut finished_height = None;
+        let blocks = chain.blocks();
+        let bundles = chain.range().filter_map(|block_number| {
+            blocks
+                .get(&block_number)
+                .map(|block| block.hash())
+                .zip(chain.execution_outcome_at_block(block_number))
+        });
+
+        for (block_hash, outcome) in bundles {
+            #[cfg(debug_assertions)]
+            assert!(outcome.len() == 1, "should only contain single block");
+            let block_number = outcome.first_block();
+            let state_diff = outcome.bundle.into();
+
+            // TODO: maybe put db writes in another thread
+            if let Err(err) = self
+                .db
+                .put_state_diff(block_hash, block_number, &state_diff)
+            {
+                error!(?err, ?block_hash);
+                break;
+            }
+
+            finished_height = Some(BlockNumHash::new(block_number, block_hash))
+        }
+
+        Ok(finished_height)
```

Comment:
If the db operation has an `Error` then the function merrily returns an `Ok(Some(T))` instead of `Error`, is that really intended? Because the state diff is not in the DB and the function clearly succeed?

Reply:
Yeah, that's the intent. Could your expand why it can be problematic or propose a better way please?

Reply:
I feel like we don't want to lose these payloads.

Reply:
well, that's exactly what's being achieved here. We send the event `ExExEvent::FinishedHeight(finished_height)` only for the block height that has been committed to the db.